File: 3sum/solution.cpp
class Solution {
public:
    std::vector<std::vector<int>> threeSum(std::vector<int>& nums) {
        std::sort(nums.begin(), nums.end());

        std::vector<std::vector<int>> res;

        int lastFirst = nums[0]-1;
        for( size_t i = 0; i < nums.size(); ++i ) {
            if( nums[i] == lastFirst ) {
                continue;
            }
            lastFirst = nums[i];
            
            std::map<int, int> seen; // map[value_which_is_needed, second_value]
            int lastResLastValue = nums[i] - 1;
            for( size_t j = i + 1; j < nums.size(); ++j ) {
                if( nums[j] == lastResLastValue ) {
                    continue;
                }

                auto it = seen.find(nums[j]);
                if( it == seen.end() ) {
                    seen[-nums[i]-nums[j]] = nums[j];
                    continue;
                }
                res.push_back({nums[i], seen[nums[j]], nums[j]});
                lastResLastValue = nums[j];
            }
        }
        return res;
    }
};

File: 3sum/solution.rs
// This one was horrible. I tried to get way too clever at first. Confusing!
use std::cmp::Ordering;
impl Solution {
    pub fn three_sum(mut nums: Vec<i32>) -> Vec<Vec<i32>> {
        nums.sort();

        let mut res: Vec<Vec<i32>> = Vec::with_capacity(nums.len()/3 + 1);

        for (i, &x) in nums.iter().enumerate() {
            let mut a = i + 1;
            let mut b = nums.len() - 1;
            if 0 < i && nums[i-1] == nums[i] {
                continue
            }

            while a < b {
                match (nums[a] + nums[b]).partial_cmp(&-x).unwrap() {
                    Ordering::Equal => {
                        let next = vec![x, nums[a], nums[b]];
                        while a < b && nums[a] == next[1] { a += 1 }
                        while a < b && nums[b] == next[2] { b -= 1 }
                        res.push(next);
                    },
                    Ordering::Less => a += 1,
                    Ordering::Greater => b -= 1,
                }
            }
        }

        res
    }
}
